### PR TITLE
Fixes #25485 - Release name required for debian

### PR DESCRIPTION
**Pull request by sbernhard**

### CODE REVIEW

File: app/models/operatingsystem.rb [hunk 1]
```diff
@@ -36,6 +36,7 @@ class Operatingsystem < ApplicationRecord
             :uniqueness => { :scope => [:major, :minor], :message => N_("Operating system version already exists")}
   validates :description, :uniqueness => true, :allow_blank => true
   validates :password_hash, :inclusion => { :in => PasswordCrypt::ALGORITHMS }
+  validates :release_name, :presence => true, :if => Proc.new { |o| o.family == 'Debian' }
```
Comment (tbrisker) [inline app/models/operatingsystem.rb]: Shouldn't this be done in `app/models/operatinsystems/debian.rb`?

**Review by tbrisker — CHANGES_REQUESTED** [review]

**Comment by sbernhard** [comment]
Would be good but unfortunately, the "validates" in operatingsystem/debian.rb are not executed.

**Comment by tbrisker** [comment]
@sbernhard - Ah, right, when we create an operating system we don't have the subclass set yet. 
However, i noticed `:set_family` callback is executed in `before_save` (which runs after validation), could there be a case in which the family isn't set yet when the validation runs? maybe we need to move that to `before_validate`?

**Comment by sbernhard** [comment]
The test "release_name should be nil when lsbdistcodename isn't set on Debian" is broken (as expected). Does someone has a good idea how to proceed? What should be done, in case a debian-based system doesn't provide the lsbdistcodename (=release_name) during the fact import?

(looks like this is the last issue which breaks the tests: https://ci.theforeman.org/job/test_develop_pr_core/20831/ )

**Comment by tbrisker** [comment]
@sbernhard I am not very familiar with the workings of debian systems, but is that a valid scenario (i.e. a debian not reporting distname in facts)? if not - than that test isn't valid and can be removed. if it is - we need to decide what we do in that case - use some sort of placeholder ("unknown"?), try to guess, fail the import?

**Comment by sbernhard** [comment]
[test katello]

**Comment by sbernhard** [comment]
Looks like the katello test are broken but don't know why and I don't think its because this PR. Any suggestions what to do?

**Comment by tbrisker** [comment]
@sbernhard don't worry about the katello failure, it is unrelated. I assume from the latest commit that the dist name is always reported on all debians?

**Comment by sbernhard** [comment]
Thanks @tbrisker for the information. 

It looks like that the release name should be always in the facts. Don't know why it shouldn't be there.

**Comment by tbrisker** [comment]
@sbernhard i did a bit more research and turns out that this fact only works on debians if you have the `lsb-release` package installed (see https://blog.x-way.org/Linux/2012/02/14/Fix-empty-puppet-lsbdistcodename-on-Debian.html for example). So I think to make sure we don't break fact imports with this change we need to add some handling for it around https://github.com/theforeman/foreman/blob/develop/app/services/puppet_fact_parser.rb#L40 - perhaps set to 'unknown' in case it wasn't set before and isn't in the facts (so we don't update an existing one)

**Comment by sbernhard** [comment]
> @sbernhard i did a bit more research and turns out that this fact only works on debians if you have the `lsb-release` package installed (see https://blog.x-way.org/Linux/2012/02/14/Fix-empty-puppet-lsbdistcodename-on-Debian.html for example). So I think to make sure we don't break fact imports with this change we need to add some handling for it around https://github.com/theforeman/foreman/blob/develop/app/services/puppet_fact_parser.rb#L40 - perhaps set to 'unknown' in case it wasn't set before and isn't in the facts (so we don't update an existing one)

Thank you very much for the help @tbrisker.

**Comment by sbernhard** [comment]
> @sbernhard i did a bit more research and turns out that this fact only works on debians if you have the `lsb-release` package installed (see https://blog.x-way.org/Linux/2012/02/14/Fix-empty-puppet-lsbdistcodename-on-Debian.html for example). So I think to make sure we don't break fact imports with this change we need to add some handling for it around https://github.com/theforeman/foreman/blob/develop/app/services/puppet_fact_parser.rb#L40 - perhaps set to 'unknown' in case it wasn't set before and isn't in the facts (so we don't update an existing one)

Thank you very much for your help @tbrisker. Added one more test to represent the change.

**Comment by tbrisker** [comment]
Thanks @sbernhard, one last request regarding this - if there are any debian OSes in the db already without release name this will make them invalid, can you add a migration updating them? Something along the lines of 
```ruby
OperatingSystem.where(type: 'Debian', release_name: [nil, '']).update_all(release_name: 'unknown')
```
should do it.

**Comment by jlsherrill** [comment]
[test foreman]
[test katello]

**Review by tbrisker — APPROVED** [review]
Thanks @sbernhard !

**Pull request merged**